### ISSUE TITLE
New Failure: bm_fullstack_streaming_pump --benchmark_filter=BM_PumpStreamServerToClient<MinSockPair>$

### ISSUE THREAD

**Issue by dgquintas** [grpc/grpc#13429]
- Test: bm_fullstack_streaming_pump --benchmark_filter=BM_PumpStreamServerToClient<MinSockPair>$
- Poll Strategy: epollex
- URL: https://kokoro2.corp.google.com/job/grpc/job/ubuntu/job/master/job/grpc_basictests_c_cpp_opt/164

**Comment by dgquintas** [comment]
Seems to be the same one, but now with an asan trace, heap-use-after free: https://sponge.corp.google.com/target?id=ccdc5d62-9302-4523-9c2d-edd0ec1bec76&target=github/grpc/c%2B%2B_linux_asan_native&searchFor=&show=ALL&sortBy=STATUS

**Comment by jtattermusch** [comment]
This is failing 16% of all asan builds (because the test case runs multiple times in each suite which amplifies the flakiness rate).

https://sponge.corp.google.com/invocation?id=795838cc-15e1-4dee-a072-7ccbf6e8307c
https://sponge.corp.google.com/invocation?id=dcafccf1-2899-475d-8474-2683a9c57c9e
https://sponge.corp.google.com/invocation?id=676654ff-6a73-417d-ace8-e4dc7a5a9fad
https://sponge.corp.google.com/invocation?id=6f3bb451-9e4d-4b45-9fae-a0121a02036c
https://sponge.corp.google.com/invocation?id=59ec1149-8061-4137-b183-19cd23a9a2da

**Comment by mehrdada** [comment]
Failure: json_run_localhost:cpp_protobuf_async_unary_qps_unconstrained_2waysharedcq_insecure

https://sponge.corp.google.com/target?id=f5192b51-10fe-484b-91f2-743d20d3fc75&target=github/grpc/aggregate_tests&searchFor=&show=ALL&sortBy=STATUS

**Comment by jtattermusch** [comment]
More data:
- only happening with `epollex`
- test is timing out (and the preset timeout is actually too high - 1200 or 3600sec).

TSAN:
https://sponge.corp.google.com/invocation?id=7bddf59f-9c48-4a72-93f2-1dcb232e8404

ASAN:
https://sponge.corp.google.com/invocation?id=57c3dff8-b7ab-4d9e-af02-9f56ea93620d

dbg:
https://sponge.corp.google.com/invocation?id=404c1aa9-2355-485d-ac19-74fa1c6f4536

opt:
https://sponge.corp.google.com/invocation?id=04ea9c1f-5688-4fa2-b2fe-d509e33994e0

**Comment by jtattermusch** [comment]
I'm able to reproduce on a VM with
`tools/run_tests/run_tests.py -l c++ -c opt -r 'bm_fullstack_streaming_pump.*SockPair' -n 50 --force_use_pollers epollex` (given enough runs).

The test gets stuck in https://github.com/grpc/grpc/blob/123547c9625c56fdf5cb4ddd1df55ae0c785fa60/test/cpp/microbenchmarks/fullstack_fixtures.h#L211 (I'm not yet sure why).

**Comment by dgquintas** [comment]
@ncteisen given all this info, do you think it'd be possible to tackle this issue soon? It's responsible for 16% of non-green PRs.

**Comment by ncteisen** [comment]
@sreecha is going to take a look and see if his #14217 will fix this. If not, I will dive in. I need to update my machine to enable epollex and then I will start digging

**Comment by sreecha** [comment]
Issue #14217 does not seem to fix this. However, the tests also fail for `poll` polling engine. 
Note that the test is disabled for `poll` and `poll-cv` (Just to reduce the number of polling engines this runs on; no other particular reason for disabling).

So I had to modify `build.yaml` to enable the tests for `poll` engine. 

Like Jan mentioned earlier, this is stuck in server shutdown path. The thread is waiting for server shutdown to complete. It is not clear why the shutdown isn't progressing.. I suspect that progress isn't happening on some fd. 

I will try to add more detailed tracing..

**Comment by sreecha** [comment]
I looked at this some more.. Don't have the root cause yet but the server is stuck in shutdown because it is waiting for the channel to be destroyed.

The server shutdown method (i.e https://github.com/grpc/grpc/blob/v1.8.6/src/core/lib/surface/server.cc#L1202) 
  - Kills pending work: https://github.com/grpc/grpc/blob/v1.8.6/src/core/lib/surface/server.cc#L1247 (which just fails any pending calls)
-  Sees if the shutdown can complete (see: https://github.com/grpc/grpc/blob/v1.8.6/src/core/lib/surface/server.cc#L680)
-  The shutdown does not happen if the listeners and channels are not destroyed (once all of them are shutdown, the shutdown would proceed - i.e `maybe_finish_shutdown` is also calledwhen channel is destroyed, when listener is destroyed ).  https://github.com/grpc/grpc/blob/v1.8.6/src/core/lib/surface/server.cc#L691

When I looked at the failing test (i.e test where the thread is stuck in shutdown), I notice that the server does not have any listeners and it has one channel that is in idle state (see `server-> root_channel_data`)..not sure why.

**Comment by jtattermusch** [comment]
Thanks for the updates, @sreecha. Btw, after fixing other stuff, this is now our most frequent flake, so it would be beautiful if we could fix this.

**Comment by ncteisen** [comment]
I have set aside some time today to work on this

**Comment by jtattermusch** [comment]
Any progress here?

**Comment by sreecha** [comment]
I looked at it some more last evening ..no update yet. I'll send an update by EOD

**Comment by sreecha** [comment]
I found the root cause for the issue. 

TLDR; Server shutdown does a write that **sometimes** does not complete and needs some thread to do a cq-next() to complete the write. This thread is single threaded and the thread is blocked in the server shutdown call and no other thread is doing a cq-next.

The fix may look trivial but it is not. The problem is a bit more subtle. Here are more details of the issue:
1) During teardown, the fixture first calls the following operations (in that order - all in one thread)
 - Server shutdown
 - Cq shutdown
 - Drain Cq 
(source [here](https://github.com/grpc/grpc/blob/v1.9.0/test/cpp/microbenchmarks/fullstack_fixtures.h#L86))

2) Server Shutdown is a **blocking call**  - It internally calls the grpc API `grpc_server_shutdown_and_notify` and **blocks** on a separate shutdown_cq for to be notified of the shutdown completion. (see source [here](https://github.com/grpc/grpc/blob/v1.9.0/src/cpp/server/server_cc.cc#L557))

3) `grpc_server_shutdown_and_notify` as a part of shutting down sends a GOAWAY on all its connections (i.e a write operation) and once the GOAWAY write is done, it closes the connection (i.e 'transport'). Unless the transport is closed, it won't release the reference to the server and the shutdown will not complete.

4) Now, in most of the cases the GOAWAY write finishes just fine in a synchronous way. In some cases the `sendmsg` API may written `EAGAIN` (see [here](https://github.com/grpc/grpc/blob/v1.9.0/src/core/lib/iomgr/tcp_posix.cc#L564))

5) When this happens, the write happens only when some thread calls `grpc_completion_queue_next()` on the server completion queue.
(But from step (1) above, that does not happen until AFTER the server and cq shutdown) and hence the thread blocks indefinitely on server shutdown.

In most of the cases, in step (4), `sendmsg()` succeeds without sending EAGAIN and everything works out fine.

---
I will fix the test for now by calling C-Core APIs directly (i.e `grpc_shutdown_and_notify()` and pass the same server cq as the shutdown cq) . I can't think of a better way at this point.

In future, it would really help to add a Async server Shutdown at C++ level.

**Comment by dgquintas** [comment]
Sree you are THE BOSS!

**Comment by sreecha** [comment]
Fix here: https://github.com/grpc/grpc/pull/14351

**Comment by yang-g** [comment]
Approved the fix. I think it is fair to say the user of async server should keep polling the ServerCqs until the server is shut down. And thus the real applications should not have this problem.

**Comment by sreecha** [comment]
@yang-g , yes.. you are right - we are technically not violating our contract and this is not a "bug" in grpc code. However, I am thinking it might help if we exposed a  `ShutdownAsync()` API at C++ level. This way, user can pass their own completion queue (which can be one of the server completion queues) and even a single thread can coordinate the shutdown.
